dali/pipeline/operators/reader/coco_reader_op.cc

```diff
@@ -14,47 +14,441 @@
+      if (0 == strcmp(internal_key, "id")) {
+        id = parser.GetInt();
+      } else {
+        parser.SkipValue();
+      }
+    }
+    DALI_ENFORCE(id != -1, "Missing category ID in the JSON annotations file");
+    category_ids.insert(std::make_pair(id, new_id));
+    new_id++;
+  }
+}
+
+void parse_annotations(
+  LookaheadParser &parser, std::vector<Annotation> &annotations, float size_threshold, bool ltrb) {
+  RAPIDJSON_ASSERT(parser.PeekType() == kArrayType);
+  parser.EnterArray();
+  while (parser.NextArrayValue()) {
+    detail::Annotation annotation;
+    if (parser.PeekType() != kObjectType) {
+      continue;
+    }
+    parser.EnterObject();
+    while (const char* internal_key = parser.NextObjectKey()) {
+      if (0 == strcmp(internal_key, "image_id")) {
+        annotation.image_id_ = parser.GetInt();
+      } else if (0 == strcmp(internal_key, "category_id")) {
+        annotation.category_id_ = parser.GetInt();
+      } else if (0 == strcmp(internal_key, "bbox")) {
+        RAPIDJSON_ASSERT(parser.PeekType() == kArrayType);
+        parser.EnterArray();
+        int i = 0;
+        while (parser.NextArrayValue()) {
+          annotation.box_[i] = parser.GetDouble();
+          ++i;
+        }
+      } else {
+        parser.SkipValue();
+      }
+    }
+    if (!annotation.FitsUnder(size_threshold)) {
+      continue;
+    }
+    if (ltrb) {
+      annotation.ToLtrb();
+    }
+    annotations.emplace_back(std::move(annotation));
+  }
+}
+
+void parse_json_file(
+  const OpSpec &spec,
+  std::vector<detail::ImageInfo> &image_infos,
+  std::vector<detail::Annotation> &annotations,
+  std::map<int, int> &category_ids) {
+  const auto annotations_file = spec.GetArgument<string>("annotations_file");
+  std::ifstream f(annotations_file);
+  DALI_ENFORCE(f, "Could not open JSON annotations file");
+  f.seekg(0, std::ios::end);
+  size_t file_size = f.tellg();
+  std::unique_ptr<char, std::function<void(char*)>> buff(
+    new char[file_size],
+    [](char* data) {delete [] data;});
+  f.seekg(0, std::ios::beg);
+  f.read(buff.get(), file_size);
+
+  detail::LookaheadParser parser(buff.get());
+
+  RAPIDJSON_ASSERT(parser.PeekType() == kObjectType);
+  parser.EnterObject();
+  while (const char* key = parser.NextObjectKey()) {
+    if (0 == strcmp(key, "images")) {
+      detail::parse_image_infos(parser, image_infos);
+    } else if (0 == strcmp(key, "categories")) {
+      detail::parse_categories(parser, category_ids);
+    } else if (0 == strcmp(key, "annotations")) {
+      parse_annotations(
+        parser,
+        annotations,
+        spec.GetArgument<float>("size_threshold"),
+        spec.GetArgument<bool>("ltrb"));
+    } else {
+      parser.SkipValue();
+    }
+  }
+  f.close();
+}
+
+}  // namespace detail
+
+
+void COCOReader::DumpMetaFiles(const std::string path, const ImageIdPairs &image_id_pairs) {
+  detail::dump_meta_file(
+    offsets_,
+    path + "offsets.txt");
```

Comment:
Do we want to leave up to the caller to end path with `/`?
Or have is `path` is non empty:
```
path + "/offsets.txt");
```

Reply:
Done